### ISSUE TITLE
Try and handle token type embeddings more elegantly

### ISSUE THREAD

**Issue by dpressel** [dpressel/mead-baseline#536]
For simple cases, we dont actually use token embeddings, but allocating this as a `zeros_like` seems pretty wasteful.

There are basically 2 use-cases that seem to make sense

1. we dont care about token embeddings and so we really want the 0-weight from them.  This is just a bias then that can be added to the input embeddings (word_embeddings + pos_embeddings + bias`)
2. we do care about the token embeddings, which suggests that we could provide a vectorizer for each of the fields, so having them as 2 fields makes sense -- one for word and one for tokens (we can always get the positions internally).  In fact even if we wanted to provide all 3 as vectorizers, it should work with our `EmbeddingsStack` except that the `EmbeddingsStack` currently has no reduction operator.  This seems quite useful and straightforward to add as its a generalization of the current concat behavior.  The user could then indicate in the `model` block of the config what `embeddings_reduction` to use (and the default would remain `concat`).  This could also produce some interesting capabilities we havent yet been able to try as a user could provide custom reductions for custom use-cases.

**Comment by dpressel** [comment]
Its possible to change the code a bit which makes this easier.  RN, the `LayerNorm` is typically executed afterward in the `TransformerEncoderStack` but when we are doign `layer_norms_after` we then make it the `Identity`.  The `Embeddings` then necessarily have to handle the LN but if we moved the LN up here, then we could potentially get around this which could allow us to move the LN outside of the `Embeddings`.  Then we can compose our features by having a TT-vectorizer and a feature vectorizer that maps to the Pos+Token embeddings.  This would prevent us from having to have a separate class with LN for embeddings

**Comment by dpressel** [comment]
Here is the current BERT embedding object:

```

class BERTLookupTableEmbeddings(LookupTableEmbeddings):
    """
    BERT style embeddings with a 0 token type

    TODO: Get rid of this, we dont need it anymore
    If you want to use BERT with token types, make a LearnedPositionalLookupTableEmbeddings feature
    and a LookupTableEmbeddings feature (for the token type)
    and put them in an EmbedStack with an embeddings_reduction='sum' on the model
    And if you dont want that, use the LearnedPositionalLookupTableEmbeddingsWithBias, which
    will add the BERT token_type=0 weights into the pos + word_embed and is more efficient
    than this class, since it doesnt do any memory allocation on the fly
    """
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.dropout = nn.Dropout(kwargs.get('dropout', 0.1))
        self.mxlen = int(kwargs.get('mxlen', 512))
        self.tok_type_vsz = kwargs['tok_type_vsz']
        self.pos_embeddings = nn.Embedding(self.mxlen, self.get_dsz())
        self.tok_embeddings = nn.Embedding(self.tok_type_vsz, self.get_dsz())
        self.ln = nn.LayerNorm(self.get_dsz(), eps=1e-12)

    def forward(self, x):
        zeros = torch.zeros_like(x)
        x = super().forward(x)
        x = x + self.positional(x.size(1)) + self.tok_embeddings(zeros)
        x = self.ln(x)
        return self.dropout(x)

    def positional(self, length):
        return self.pos_embeddings(
            torch.arange(length, dtype=torch.long, device=self.pos_embeddings.weight.device)
        ).unsqueeze(0)
```

An observation that can simplify this drastically is that the `EmbeddingsStack` object already supports a `reduction` operator and does a dropout on the final result.

If we could provide an `add-layer-norm` `reduction`, we could support the token type embeddings compositionally:

```
EmbeddingsStack({'x': LearnedPositionalLookupTableEmbeddings(...), 'tt': PositionalLookupTableEmbeddings, dropout=kwargs.get('embeddings_dropout'), reduction='add-layer-norm')
```

**Comment by dpressel** [comment]
This is just waiting on checkpoint generation to close
